packages/myst-cli/src/project/fromToc.ts

```diff
@@ -16,39 +23,96 @@ import type {
   PageSlugs,
 } from './types.js';
 
-function pagesFromChapters(
+export const tocFile = (filename: string): string => {
+  if (extname(filename) === '.yml') return filename;
+  return join(filename, '_toc.yml');
+};
+
+export function validateTOC(session: ISession, path: string): boolean {
+  const filename = tocFile(path);
+  if (!fs.existsSync(filename)) return false;
+  const contents = fs.readFileSync(filename).toString();
+  try {
+    const { toc, didUpgrade } = parseTOC(contents);
+
+    if (didUpgrade) {
+      addWarningForFile(
+        session,
+        filename,
+        `${filename} is out of date: see https://executablebooks.org/en/latest/blog/2021-06-18-update-toc`,
+        'warn',
+        { ruleId: RuleId.validTocStructure },
+      );
+    }
+
+    return true;
+  } catch (error) {
+    const { message } = error as unknown as Error;
+    addWarningForFile(
+      session,
+      filename,
+      `Table of Contents (ToC) file did not pass validation:\n - ${message}\n - An implicit ToC will be used instead\n`,
+      'error',
+      { ruleId: RuleId.validTocStructure },
+    );
+    return false;
+  }
+}
+
+function pagesFromSubtree(
```

Comment:
To my knowledge, this function `pagesFromSubtree` is mapping from a nested structure to a flat list of page/directory types. Is this understanding correct?